text/0000-proc-macro-include.md

```diff
@@ -0,0 +1,175 @@
+- Start Date: 2021-11-24
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Proc macros can now effectively `include!` other files and process their contents.
+This both allows proc macros to communicate that they read external files,
+and to maintain spans into the external file for more useful error messages.
+
+# Motivation
+[motivation]: #motivation
+
+- `include!` and `include_str!` are no longer required to be compiler built-ins,
+  and could be implemented as proc macros.
+- Help incremental builds and build determinism, by proc macros telling rustc which files they read.
+- Improve proc macro sandboxability and cacheability, by offering a way to implement this class of
+  file-reading macros without using OS APIs directly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## For users of proc macros
+
+Nothing changes! You'll just see nicer errors and fewer rebuilds
+from procedural macros which read external files.
+
+## For writers of proc macros
+
+Three new functions are provided in the `proc_macro` interface crate:
+
+```rust
+/// Read the contents of a file as a `TokenStream` and add it to build dependency info.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// May fail for a number of reasons, for example, if the string contains unbalanced delimiters
+/// or characters not existing in the language.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// Source spans are constructed for the read file. If you use the spans of this token stream,
+/// any resulting errors will correctly point at the tokens in the read file.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include<P: AsRef<str>>(path: P) -> Result<TokenStream, std::io::Error>;
+
+/// Read the contents of a file as a string and add it to build dependency info.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include_str<P: AsRef<str>>(path: P) -> Result<String, std::io::Error>;
```

Comment:
You may be interested in https://github.com/rust-lang/rust/pull/63681 which discussed whether the strings needs to be included as a binary (exactly), or as source code (with some light normalization).

Reply:
Especially since this gives a chance to post-process the sting, it definitely makes sense to include the string as-is.

Reply:
Note that `fn include` should load the file as source code and do the lexer preprocessing anyway.